### PR TITLE
Move compiler test ref assemblies to NuPkg

### PR DESCRIPTION
**Pull request by jaredpar**
The compiler and ide tests depend on in memory reference assemblies for constructing `Compilation` objects. These in memory reference assemblies are not offered by the platform and hence we were just generating them as a part of our build. That had a number of limitations though:

1. The reference assemblies for .NET Core are tied to the version of `Microsoft.NETCore.App.Ref.nupkg`. That is not a side effect free NuPkg. Referencing it has other impacts to the build. Particularly it is very difficult for us to build referencing multiple versions of that package hence we can't easily have tests that rely on a mix of `net5.0` and `net6.0`. That is very important for the compiler team though when working on cross cutting features.
2. The build infra puts all of the ref assemblies into a single project that is then referenced by every single test project we have. That means every test project carries a copy of the ref assemblies whether or not we use them. There is no realistic way to have a project reference only a subset of the ref assemblies. While the negative impact of build output size is often over stated this is one case where it has caused us real problems in the past (caused us to go over the 10gig limit for azure machines a few times). 

This change begins the process of moving us from generating the ref assemblies in our build to just referencing them as NuPkg from https://github.com/jaredpar/basic-reference-assemblies. The code for generating those assemblies is effectively what we had in our repo. I just cleaned it up a bit and changed it to generate assemblies per TFM so that it was easier to consume them on a "need" basis. Also because that build is much more disconnected it's easy to generate every TFM simultaneously and let us reference `net5.0` and `net6.0` at the same time.

### CODE REVIEW

**Comment by jaredpar** [comment]
@dotnet/roslyn-compiler PTAL

File: src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj [hunk 1]
```diff
@@ -96,11 +96,13 @@
...
     <PackageReference Include="Microsoft.NET.Build.Extensions" Version="$(MicrosoftNETBuildExtensionsVersion)" IncludeAssets="none" PrivateAssets="all" GeneratePathProperty="true" />
-    <PackageReference Include="Microsoft.NETCore.App.Ref" Version="$(MicrosoftNETCoreAppRefVersion)" IncludeAssets="none" PrivateAssets="all" GeneratePathProperty="true" />
     <PackageReference Include="jnm2.ReferenceAssemblies.net35" Version="$(jnm2ReferenceAssembliesnet35Version)" IncludeAssets="none" PrivateAssets="all" GeneratePathProperty="true" />
     <PackageReference Include="System.Threading.Tasks.Extensions" Version="$(SystemThreadingTasksExtensionsVersion)" IncludeAssets="none" PrivateAssets="all" GeneratePathProperty="true" />
     <PackageReference Include="Microsoft.CSharp" Version="$(MicrosoftCSharpVersion)" GeneratePathProperty="true" />
     <PackageReference Include="Microsoft.VisualBasic" Version="$(MicrosoftVisualBasicVersion)" IncludeAssets="none" PrivateAssets="all" GeneratePathProperty="true" />
+
+    <PackageReference Include="Basic.Reference.Assemblies.NetStandard20" Version="$(BasicReferenceAssembliesNetStandard20Version)" />
```
Comment (AlekseyTs) [inline src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj:104]: >Basic.Reference.Assemblies.NetStandard20 [](http://example.com/codeflow?start=31&length=40)

How this package is created?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

File: src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj:104]: For every major target framework there is a NuPkg that contains all of the relevant reference assemblies. For example: 

- [netstandard2.0](https://www.nuget.org/packages/NETStandard.Library/2.0.3)
- [net461](https://www.nuget.org/packages/Microsoft.NETFramework.ReferenceAssemblies.net461)

The NuPkg all have a very similar structure for how the ref assemblies are laid out. The basic assemblies repo has a [script](https://github.com/jaredpar/basic-reference-assemblies/blob/main/Scripts/Generate.ps1) that just walks that directory structure and generates targets + code that will embed the ref assemblies as resources and expose them as named members on types. 

It's very similar to the way we generate them in our repo but just a bit more polished and generalized.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

File: src/Compilers/Test/Core/Generated.cs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (333fred) [inline src/Compilers/Test/Core/Generated.cs:1]: File encoding change?

**Review by 333fred — APPROVED** [review]
One question on file encoding, otherwise LGTM

File: src/Compilers/Test/Core/TargetFrameworkUtil.cs [hunk 3]
```diff
@@ -71,9 +77,75 @@ public enum TargetFramework
         MinimalAsync,
     }
 
+    /// <summary>
+    /// This type holds the reference information for the latest .NET Core platform. Tests
+    /// targeting .NET core specifically should use the references here. As the platform moves
+    /// forward these will be moved to target the latset .NET Core supported by the compiler
```
Comment (cston) [inline src/Compilers/Test/Core/TargetFrameworkUtil.cs:83]:
```suggestion
    /// forward these will be moved to target the latest .NET Core supported by the compiler
```

File: src/Compilers/Test/Core/TargetFrameworkUtil.cs [hunk 4]
```diff
@@ -71,9 +77,75 @@ public enum TargetFramework
...
+    }
+
+    /// <summary>
+    /// This type holds the reference information for the latest .NET Framework. These should be
+    /// used by tests that are specific to .NET Framework. This moves forward much more rarely but
+    /// when it does move forward these will change
+    /// </summary>
+    public static class NetFramework
+    {
+        public static ImmutableArray<MetadataReference> StandardReferences => ImmutableArray.Create<MetadataReference>(
```
Comment (cston) [inline src/Compilers/Test/Core/TargetFrameworkUtil.cs:125]: >static [](http://example.com/codeflow?start=15&length=6)

`readonly`

File: src/Compilers/Test/Core/TargetFrameworkUtil.cs [hunk 5]
```diff
@@ -71,9 +77,75 @@ public enum TargetFramework
...
+            Net50.mscorlib,
+            Net50.SystemRuntime,
+            Net50.SystemCore,
+            Net50.SystemConsole,
+            Net50.SystemLinq,
+            Net50.SystemLinqExpressions,
+            Net50.SystemThreadingTasks,
+            Net50.SystemCollections);
+
+        public static PortableExecutableReference netstandard = Net50.netstandard;
```
Comment (cston) [inline src/Compilers/Test/Core/TargetFrameworkUtil.cs:104]: >static [](http://example.com/codeflow?start=15&length=6)

`readonly` for all of these fields.

**Review by cston — APPROVED** [review]
Minor comments only.

File: src/Compilers/Test/Core/TargetFrameworkUtil.cs (hunk 3 above)
Comment (jaredpar) [reply src/Compilers/Test/Core/TargetFrameworkUtil.cs:83]: I have no excuse

**Pull request merged**